Add argument 'launch'

trwyant opened the pull request:
If specified, this argument attempts to launch a browser on the
debugger's port. The value of the argument is the name of the browser to
launch. If the argument is omitted, Browser::Open is loaded, and its
favored browser is used. If the attempt to load Browser::Open fails, an
exception is raised.

trwyant committed:
Add argument 'launch'

If specified, this argument attempts to launch a browser on the
debugger's port. The value of the argument is the name of the browser to
launch. If the argument is omitted, Browser::Open is loaded, and its
favored browser is used. If the attempt to load Browser::Open fails, an
exception is raised.

brummett commented on lib/Devel/hdb.pm:
```diff
@@ -33,7 +33,12 @@ sub import {
-        }
+        } elsif ($param =~ m/\Alaunch(:(.+))?/) {
+	    our $LAUNCH = $1 ? $2 : do {
+		require Browser::Open;
```
If the require fails (say, because ```Browser::Open``` isn't installed), it should print something to that affect to STDERR with some instructions on how to get it installed.

trwyant replied:
Good thought. I will look into it.

brummett commented on lib/Devel/hdb/App.pm:
```diff
@@ -151,6 +152,9 @@ sub init_debugger {
     local $@;
     eval { $self->load_settings_from_file() };
 
+    system { $Devel::hdb::LAUNCH } $Devel::hdb::LAUNCH, $self->_gui_url()
```
If this ```system()``` fails to start the requested program, it should print an error including ```$!``` so the user can figure out what went wrong.

trwyant replied:
Also a good thought, but I am not sure that $! is the correct variable. $! is Perl's representation of the C errnum/errstr stuff. I think things like IPC::System::Simple throw an exception if the system() fails. But I will try to come up with something.

brummett replied:
If you use ```$!``` in string context, it'll stringify errno.  For the easy case where you specify a program that doesn't exist, then the ```system()``` will fail.  ```$?``` is -1 and ```$!``` is "No such file or directory".  Other errors like the internal fork() failing will stringify similarly.

brummett reviewed (changes requested):
Thanks for the submission!

I have a couple of suggestions that'll make it easier for users to figure out the problems if something goes wrong.

trwyant committed:
Add warnings if browser launch fails.

This is actually two warnings: one for failure to load Browser::Open if
the command is defaulted, and one if the system() call fails for some
reason.

trwyant commented:
After thinking about it, I decided that in both cases the appropriate response was a warning. The logic is that if 'launch' fails for any reason, the user response is the same as if it was not specified, so it was overkill to throw an exception. I decided against reporting in $@ or $! because I was not sure the user would actually see this. But it's your package, and you may feel different.

I have been able to cover the case where Browser::Open can not be loaded, and the case where the system() command fails. What I am unable to figure out is how to do anything if the system command succeeds but does not launch a browser. For example, under U*ix, '-d:hdb=launch:true'.

I also have great faith in my ability to write unclear prose. Feel free to wordsmith the warnings.

brummett commented:
> After thinking about it, I decided that in both cases the appropriate response was a warning. The logic is that if 'launch' fails for any reason, the user response is the same as if it was not specified, so it was overkill to throw an exception. I decided against reporting in $@ or $! because I was not sure the user would actually see this. But it's your package, and you may feel different.

I think it would be appropriate to stop if the launch failed because it's not able to do the thing you asked it to do.  Reporting that fact and some clue as to why it failed is going to help the user when something goes wrong, instead of leaving them wondering why it didn't work.

If you're worried that the user wouldn't see it, then a test is in order.  Try launching it for a working case (using, say, /bin/true or /bin/sleep) and several failure cases (file not found, /bin/false, others?) and make sure the error shows up on STDERR where the user would normally see it.

> I have been able to cover the case where Browser::Open can not be loaded, and the case where the system() command fails. What I am unable to figure out is how to do anything if the system command succeeds but does not launch a browser. For example, under U*ix, '-d:hdb=launch:true'.

I think that much is fine.  If the user asks to run a particular program, and ```system()``` is able to launch it successfully, then that's all you can reasonably do.  It would be extraordinary measures to babysit the launched program afterward and make sure it's interacting with the debugger.

brummett commented:
The other thing I just thought of is: doesn't ```system()``` block until the launched program finishes?  If it's starting a browser or some other long-lived program, then the debugger is going to be blocked waiting instead of continuing on and debugging the program.

brummett commented:
There may also be a race condition.  If you look at ```sub _make_listen_socket``` in lib/Devel/hdb/App.pm, you're launching the browser at the end of ```init_debugger``` on line 72, but the listen socket isn't created until the call to ```Devel::hdb::Server->new()``` returns on line 91.

trwyant commented:
"Die on error" was obviously not my first thought, but it is a reasonable one, and I will update the pull request to do this. The messages will obviously have to be word-smithed.

The synchronous-system() thing needs at least a little research. It is not a problem in what I think of as the most likely cases ("*-open" under Linux, "open" under macOS, "start" under Windows). But it will become a problem the first time someone does "launch:firefox" (as opposed to "launch:firefox&" which should be well-behaved but is system-dependent and requires shelling out). I will dig into this.

As for the race condition, thank you for pointing it out. I have not had any trouble so far, but races are to be eliminated.

So I plan to address all three points, I just don't know when I will get to all of them.

trwyant committed:
Die on 'launch' error.

trwyant commented:
Well, after a long delay (involving a disk disaster among other things) I have looked into the possible race. It appears to me that the socket is still not there on line 91 -- at least, a call to $self->_get_url() returns '/debugger-gui'. What I think is happening is that init_debugger() is not actually called at line 71, but is wrapped in a closure and ultimately passed to PSGI, which calls it from PSGI::setup_listener(). Thinking about it, I seem to remember struggling a bit to find a place where Devel::hdb::App actually had all the information that was needed.

Now, this still does not mean there is no race, but it does mean I will have to dive into PSGI to figure it out.

The second point I understand to be outstanding is the launching of a browser using a command that does not, in effect, "amp off" the launch. Without adding a dependency it appears at the moment that I will need to fork(). Now, the comments make it look like this mechanism is already in use, but the only explicit call to fork() I have found is in t/HdbHelper.pm.

trwyant commented:
Quick follow-up. As I read HTTP::Server::PSGI, the {server_ready} callback is called at line 107 as the last line of setup_listener(), _after_ the port is created at line 103. So I conclude no race, though I have been wrong before.
